git_test.go

```diff
@@ -1,35 +1,476 @@
 package main
 
 import (
+	"fmt"
+	"io"
+	"log"
 	"os"
+	"path"
 	"testing"
+	"time"
 
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
 	"github.com/stretchr/testify/assert"
 	"go.opentelemetry.io/otel/attribute"
 )
 
-func TestGit(t *testing.T) {
-	workingDir, err := os.Getwd()
+var workingDir string
+var tempDir string
+
+func init() {
+	wd, err := os.Getwd()
 	if err != nil {
-		t.Error()
+		log.Fatalln("Cannot get current working dir, which is needed by tests")
 	}
 
-	scm := &GitScm{
-		repositoryPath: workingDir,
+	workingDir = wd
+}
+
+// FakeGitRepo downloads Octocat's hello-world repository from Github, providing a simple DSL to add/remove files and commit them into
+// the fake git repository
+type FakeGitRepo struct {
+	repo             *git.Repository
+	divergesFromBase bool
+	repoPath         string
+	t                *testing.T
+}
+
+func NewFakeGitRepo(t *testing.T) *FakeGitRepo {
+	cloneURL := "https://github.com/octocat/hello-world"
```

Comment:
Using a very simple git repo to add/delete files, commit, calculate logs in the tests

Reply:
An improvement could be caching the repo once, and reuse it across different tests